substrate/frame/staking/src/pallet/mod.rs

```diff
@@ -1980,6 +1983,115 @@ pub mod pallet {
 
 			Ok(Some(T::WeightInfo::deprecate_controller_batch(controllers.len() as u32)).into())
 		}
+
+		/// Restores the state of a ledger which is in an inconsistent state.
+		///
+		/// The requirements to restore a ledger are the following:
+		/// * The stash is bonded;
+		/// * If the stash has an associated ledger, its state must be inconsistent;
+		/// * If the ledger is not corrupted *but* its staking lock is out of sync.
+		///
+		/// The `maybe_*` input parameters will overwrite the corresponding data and metadata of the
+		/// ledger associated with the stash. If the input parameters are not set, the ledger will
+		/// be reset with the current values.
+		///
+		/// Upon successful execution, this extrinsic will *recreate* the ledger based on its past
+		/// state and/or the `maybe_controller` and `maybe_total` input parameters.
+		#[pallet::call_index(29)]
+		#[pallet::weight(T::WeightInfo::restore_ledger())]
+		pub fn restore_ledger(
+			origin: OriginFor<T>,
+			stash: T::AccountId,
+			maybe_controller: Option<T::AccountId>,
```

Comment:
I would for the sake of simplicity make the caller always provide all arguments. 

I am assuming that any stash that is eligible to `restore`-ed via this extrinsic is already in a "locked" state and they can change nothing, so everything can come in as an argument.

Reply:
Do you mean to also add an option to set the restored ledger's unlocking chunks? If so, I'm adding this now so we can set the ledger's chunks when restoring a ledger, if the caller wishes.